### PR TITLE
SEO: Expose advanced_seo_description post meta in the REST API (Take Two)

### PR DESCRIPTION
**Pull request by ockham**
Take Two of #11323. Alternative to #11333.
Calypso counterpart: https://github.com/Automattic/wp-calypso/pull/30033 (merged)
Brings r197556-wpcom to Jetpack
Diff: D32268-code

#### Changes proposed in this Pull Request:

**Note: following the discussion below, and in D32268-code, this PR now does not bring any new feature to Jetpack.**

* SEO: Expose advanced_seo_description post meta in the (core) REST API

#### Testing instructions:

* This should not bring any changes to Jetpack. Nothing should change.

#### Proposed changelog entry for your changes:

* SEO: Expose advanced_seo_description post meta in the REST API

### CODE REVIEW

**Comment by simison** [comment]
When testing, remember to build blocks from latest Calypso master to get SEO description in the bundle. Otherwise, availability will show up as:

```
available: false
unavailable_reason: "not_whitelisted"
```

...suppose because "seo" isn't in the JSON file listing blocks.

To build, run this in Calypso:
```
npm ci
npx lerna bootstrap --concurrency=2 --scope='@automattic/jetpack-blocks'
npx lerna run build --stream --scope='@automattic/jetpack-blocks' -- -- --output-path /path/to/jetpack/_inc/blocks/ --watch
```

Remember to update path to your Jetpack path.

**Comment by simison** [comment]
Getting PHP warning:
```
[22-Mar-2019 20:57:37 UTC] PHP Warning:  Use of undefined constant Jetpack_SEO_Posts - assumed 'Jetpack_SEO_Posts' (this will throw an Error in a future version of PHP) in /var/www/html/wp-content/plugins/jetpack/modules/seo-tools/jetpack-seo.php on line 9
[22-Mar-2019 20:57:37 UTC] PHP Stack trace:
[22-Mar-2019 20:57:37 UTC] PHP   1. {main}() /var/www/html/wp-admin/admin-ajax.php:0
[22-Mar-2019 20:57:37 UTC] PHP   2. require_once() /var/www/html/wp-admin/admin-ajax.php:22
[22-Mar-2019 20:57:37 UTC] PHP   3. require_once() /var/www/html/wp-load.php:37
[22-Mar-2019 20:57:37 UTC] PHP   4. require_once() /var/www/html/wp-config.php:84
[22-Mar-2019 20:57:37 UTC] PHP   5. do_action() /var/www/html/wp-settings.php:374
[22-Mar-2019 20:57:37 UTC] PHP   6. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[22-Mar-2019 20:57:37 UTC] PHP   7. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[22-Mar-2019 20:57:37 UTC] PHP   8. Jetpack::load_modules() /var/www/html/wp-includes/class-wp-hook.php:286
[22-Mar-2019 20:57:37 UTC] PHP   9. include_once() /var/www/html/wp-content/plugins/jetpack/class.jetpack.php:1858
[22-Mar-2019 20:57:37 UTC] PHP  10. Jetpack_SEO->__construct() /var/www/html/wp-content/plugins/jetpack/modules/seo-tools.php:3
```

File: modules/seo-tools/jetpack-seo.php
Comment (simison) [inline modules/seo-tools/jetpack-seo.php]: Getting PHP warning:
```
[22-Mar-2019 20:57:37 UTC] PHP Warning:  Use of undefined constant Jetpack_SEO_Posts - assumed 'Jetpack_SEO_Posts' (this will throw an Error in a future version of PHP) in /var/www/html/wp-content/plugins/jetpack/modules/seo-tools/jetpack-seo.php on line 9
[22-Mar-2019 20:57:37 UTC] PHP Stack trace:
[22-Mar-2019 20:57:37 UTC] PHP   1. {main}() /var/www/html/wp-admin/admin-ajax.php:0
[22-Mar-2019 20:57:37 UTC] PHP   2. require_once() /var/www/html/wp-admin/admin-ajax.php:22
[22-Mar-2019 20:57:37 UTC] PHP   3. require_once() /var/www/html/wp-load.php:37
[22-Mar-2019 20:57:37 UTC] PHP   4. require_once() /var/www/html/wp-config.php:84
[22-Mar-2019 20:57:37 UTC] PHP   5. do_action() /var/www/html/wp-settings.php:374
[22-Mar-2019 20:57:37 UTC] PHP   6. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[22-Mar-2019 20:57:37 UTC] PHP   7. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[22-Mar-2019 20:57:37 UTC] PHP   8. Jetpack::load_modules() /var/www/html/wp-includes/class-wp-hook.php:286
[22-Mar-2019 20:57:37 UTC] PHP   9. include_once() /var/www/html/wp-content/plugins/jetpack/class.jetpack.php:1858
[22-Mar-2019 20:57:37 UTC] PHP  10. Jetpack_SEO->__construct() /var/www/html/wp-content/plugins/jetpack/modules/seo-tools.php:3
```

Comment (kraftbj) [inline modules/seo-tools/jetpack-seo.php]:
```suggestion
		add_action( 'init', array( 'Jetpack_SEO_Posts', 'register_post_meta' ), 20 );
```

File: modules/seo-tools/jetpack-seo.php
Comment (kraftbj) [inline modules/seo-tools/jetpack-seo.php]:
```suggestion
		add_action( 'jetpack_register_gutenberg_extensions', array( 'Jetpack_SEO_Posts', 'register_gutenberg_extension' ) );
```

**Comment by kraftbj** [comment]
@simison FIxed what should have been causing the PHP error. I'm wrestling with building the blocks and getting them into my JN so I'm a little behind on the actual testing part.

**Comment by simison** [comment]
>I'm wrestling with building the blocks and getting them into my JN 

I'm not sure if script in JN is up to date, since we've been changing the command quite frequently and don't really expect to keep JN up to date right now while we're moving things over:

https://github.com/Automattic/jurassic.ninja/blob/11f8873ec9c57dea28e536d772c24d5587a9d44e/bin/build-gutenpack.sh#L18-L19

**Comment by ockham** [comment]
Thanks for pointing that out @simison! And yeah, it's quite possible that JN isn't up-to-date, sry 'bout that.

**Review by simison — DISMISSED** [review]
I haven't done very extensive testing yet but code looks good;

- tested with post, writing SEO description saves and shows up in meta tags at the page
- removing the SEO description brings back site's contents to the meta tag
- Disabled SEO tooling from `/wp-admin/admin.php?page=jetpack_modules` and the textarea in the sidebar was gone

Code looks good, too.

**Comment by simison** [comment]
Updated JN: https://github.com/Automattic/jurassic.ninja/pull/174

File: modules/seo-tools/jetpack-seo-posts.php
Comment (jeherve) [inline modules/seo-tools/jetpack-seo-posts.php]: Shouldn't it be named `jetpack/seo`, following how other blocks and extensions are named?

File: modules/seo-tools/jetpack-seo-posts.php
Comment (jeherve) [inline modules/seo-tools/jetpack-seo-posts.php]: The Calypso editor currently offers a metabox when editing pages as well. Shouldn't we support pages too?

File: modules/seo-tools/jetpack-seo-posts.php
Comment (jeherve) [inline modules/seo-tools/jetpack-seo-posts.php]: Should we use the `sanitize_callback` parameter to make sure the meta that is saved is clean (no HTML) and no longer than 160 characters?

**Review by jeherve — CHANGES_REQUESTED** [review]
I have a few questions, below. I was also wondering about the module itself; we had talked about making the feature accessible to everyone (free). Will this change in a different PR?

File: modules/seo-tools/jetpack-seo-posts.php
Comment (simison) [reply modules/seo-tools/jetpack-seo-posts.php]: Confirming that the previous PR worked for pages, too. Good catch! Can also be follow up PR I think.

File: modules/seo-tools/jetpack-seo-posts.php
Comment (simison) [reply modules/seo-tools/jetpack-seo-posts.php]: I think the registration function in jetpack knows how to handle this and it's in line how Gutenberg's register plugin -function looks like?

Comment (ockham) [reply modules/seo-tools/jetpack-seo-posts.php]: We seem to have adopted a convention to prefix our Gutenberg _blocks_ with `jetpack/`, but our Gutenberg _plugins_ with `jetpack-`. (Not that we've been super consistent about it :sweat_smile: )

Comment (jeherve) [reply modules/seo-tools/jetpack-seo-posts.php]: Which plugins use `jetpack-`? 

- Publicize seems to be `jetpack/publicize`
- Shortlinks is `jetpack/shortlinks`

Maybe we should stick to using `/` to avoid any confusion in the future?

Comment (simison) [reply modules/seo-tools/jetpack-seo-posts.php]: I'm indifferent how do we do now for this PR but seems like we have some janitorial work to do with plugins regardless, so I'd be okay leaving it for follow up PR:

- Move non-block integrations from `extensions/blocks` to `extensions/plugins`
- ensure all slugs are named consistently
- maaaybe refactor https://github.com/Automattic/jetpack/blob/b8717477aecd787bf7078e73a69cf1f49ada56dc/class.jetpack-gutenberg.php#L100-L112 — unless it's important for wpcom?

Comment (simison) [reply modules/seo-tools/jetpack-seo-posts.php]: and just in case it wasn't clear, I'm also up for consistency and enforcing any pattern we decide on :-)

File: modules/seo-tools/jetpack-seo-posts.php
Comment (simison) [reply modules/seo-tools/jetpack-seo-posts.php]: Why 160 characters?

**Comment by jeherve** [comment]
Moving this out of the milestone for now since this seems to be on the backburner for now.

File: modules/seo-tools/jetpack-seo-posts.php
Comment (chrisfromthelc) [reply modules/seo-tools/jetpack-seo-posts.php]: We should not drop at 160, only warn the user that they've reached the optimal length. Search engines will parse longer meta descriptions and occasionally display them longer (although rare). I don't remember what the original length was (maybe 300?), but 300 would be a good spot.

As for sanitizing out HTML, yes, please. :)

**Comment by jeherve** [comment]
> Could you also look at re-starting this one, perhaps?

Noting that this was put on hold pending a decision on this:
p1HpG7-6a7-p2

We'd need to make a decision there before we can ship this feature.

**Comment by simison** [comment]
>We'd need to make a decision there before we can ship this feature.

For Jetpack specifically — it'd be ok to ship to .com, right?

**Comment by jeherve** [comment]
> it'd be ok to ship to .com, right?

Yup, correct. My bad, I always tend to think about Jetpack first. :)

**Comment by donpark** [comment]
I don't think this patch will work as-is on WPCOM (nor multisites?) because `get_current_blog_id()` returns site ID of `public-api.wordpress.com` when `register_post_meta()` is called via `init` action.

**Comment by donpark** [comment]
WPCOM patch D32268-code is ready for review. It enables SEO description post meta for all Simple sites but UI for viewing and updating the value will appear in Classic/Gutenberg editors for only sites with Business or eCommerce plan.

**Comment by donpark** [comment]
WPCOM patch D32268-code is on-hold until a decision is made, presumably by Jetpack crew, on how best to enable the feature on WPCOM without introducing WPCOM check into Jetpack code or keeping WPCOM/Jetpack out of sync. As I see it, only way out without anyone getting hurt is if Jetpack enables SEO extension by default.

**Comment by simison** [comment]
> without introducing WPCOM check into Jetpack code

Is that an issue? There are several of those already even in Gutenberg `/extensions`. Search with `IS_WPCOM`.

**Comment by donpark** [comment]
> > without introducing WPCOM check into Jetpack code
> 
> Is that an issue? There are several of those already even in Gutenberg `/extensions`. Search with `IS_WPCOM`.

It apparently is (see D32268-code comments). If this was my beef, I'd just enable SEO description meta support everywhere and call it done. Without a service that takes the data and provides value, which anyone can implement, it's just passive data that anyone should be able to set IMO. Since it's not, I'll let you the Crew have their say.

**Comment by donpark** [comment]
D32268-code is ready for review again. WPCOM aware code is now in an existing WPCOM-only file. Actual host-independent moving parts are in a synced file, ready to be called when the Crew decides to enable it on Jetpack.

**Comment by donpark** [comment]
D32268-code was deployed to WPCOM just now. Thanks to everyone, particularly @jeherve and @simison.

**Comment by simison** [comment]
Thank you @donpark ! You rock!

**Review by jeherve — APPROVED** [review]

File: modules/seo-tools/jetpack-seo-posts.php
Comment (kraftbj) [reply modules/seo-tools/jetpack-seo-posts.php]: Is there a follow-up PR for this already?

Comment (jeherve) [reply modules/seo-tools/jetpack-seo-posts.php]: I don't think anyone worked on that yet.

**Pull request merged**
